File: web/src/lib/dataProvider/composite.ts
// buildComposite — the TechnoFunda composite score. Pure + deterministic +
// golden-testable. The TECHNICAL leg is read straight from `chart` (the same
// memoized StockChartResponse the detail chart draws), so the chart and the
// composite AGREE by construction. Provisional legs (ownership, valuation)
// still count at FULL weight per the locked "strengthen-composite-now"
// decision — only NULL legs drop out and the rest renormalize.
import type { CompositeScore, Interpretation, SubScore } from "@/lib/types";
import { FLAT_PCT, percentileOfLast, round } from "./analysis";
import { STAGE_SHORT, STAGE_TONE } from "./interpret";
import type { Fundamentals } from "./mockFundamentals";
import { seedFromSymbol } from "./mockSeries";
import type { Stage, StockChartResponse, Verdict } from "./types";

const clamp = (x: number) => Math.max(0, Math.min(100, x));

const WEIGHTS = { technical: 0.4, growth: 0.3, ownership: 0.15, valuation: 0.15 } as const;

function technicalLeg(chart: StockChartResponse): SubScore {
  const stage = chart.stage;
  if (stage == null) {
    return { value: null, provisional: false, interpretation: { headline: "No technical read", tone: "neutral", detail: "Needs 170 sessions of history." } };
  }
  const stageBase = { 1: 45, 2: 75, 3: 35, 4: 12 }[stage];
  const proximity = 0.15 * ((chart.rangePos ?? 50) - 50); // ±7.5, centred so mid-range is neutral
  const rsBonus = chart.rsNewHigh ? 12 : 0;
  const trendBonus = chart.aboveMa ? 6 : 0;
  const slopeNudge = chart.maSlopePct == null ? 0 : chart.maSlopePct > FLAT_PCT ? 4 : chart.maSlopePct < -FLAT_PCT ? -4 : 0;
  const value = round(clamp(stageBase + proximity + rsBonus + trendBonus + slopeNudge), 0);
  const rangeWord = (chart.rangePos ?? 50) >= 66 ? "near the top of its 52-week range" : (chart.rangePos ?? 50) <= 33 ? "near the low of its range" : "mid-range";
  const rsWord = chart.rsNewHigh ? "RS at a new high (leading)" : "RS not yet leading";
  return {
    value,
    provisional: false,
    interpretation: { headline: STAGE_SHORT[stage], tone: STAGE_TONE[stage], detail: `${rangeWord}; ${rsWord}.` },
  };
}

function growthLeg(fund: Fundamentals): SubScore {
  const a = fund.annual;
  const last = a[a.length - 1];
  const rev = last?.revenueYoY ?? null;
  const eps = last?.epsYoY ?? null;
  if (rev == null && eps == null) {
    return { value: null, provisional: true, interpretation: { headline: "Growth history unavailable", tone: "neutral", detail: "Not enough reported periods." } };
  }
  const revScore = clamp(50 + 1.4 * (rev ?? 0));
  const epsScore = clamp(50 + 1.0 * (eps ?? 0));
  const parts = [rev != null ? revScore : null, eps != null ? epsScore : null].filter((x): x is number => x != null);
  let base = parts.reduce((s, x) => s + x, 0) / parts.length;
  if (eps != null && eps < 0) base -= 10; // explicit negative-EPS penalty
  const value = round(clamp(base), 0);
  const prevRev = a[a.length - 2]?.revenueYoY ?? null;
  const accel = rev != null && prevRev != null ? (rev > prevRev + 1 ? "accelerating" : rev < prevRev - 1 ? "decelerating" : "steady") : "steady";
  const tone: Interpretation["tone"] = value >= 62 ? "good" : value < 45 ? "bad" : "neutral";
  const revTxt = rev == null ? "n/a" : `${rev >= 0 ? "+" : "−"}${Math.abs(rev).toFixed(0)}%`;
  const epsTxt = eps == null ? "n/a" : `${eps >= 0 ? "+" : "−"}${Math.abs(eps).toFixed(0)}%`;
  return { value, provisional: rev == null || eps == null, interpretation: { headline: `Revenue ${revTxt} YoY, EPS ${epsTxt} — ${accel}`, tone, detail: null } };
}

function ownershipLeg(symbol: string): SubScore {
  const member = seedFromSymbol(symbol) % 7 === 0; // deterministic IBD50/FFTY proxy (~1/7)
  return {
    value: member ? 78 : 46,
    provisional: true, // no real 13F feed exists yet — always provisional
    interpretation: member
      ? { headline: "In the IBD 50 proxy", tone: "good", detail: "Institutional sponsorship present; provisional until a real 13F feed lands." }
      : { headline: "Not in the IBD 50 proxy", tone: "neutral", detail: "Thin ownership signal; provisional until a real 13F feed lands." },
  };
}

function valuationLeg(fund: Fundamentals, sectorPeMedian: number | null): SubScore {
  const pePctile = percentileOfLast(fund.peSeries);
  const peNow = fund.latest.pe;
  const peVsSector = sectorPeMedian && peNow ? round(100 * (peNow / sectorPeMedian - 1), 0) : null;
  if (pePctile == null) {
    return { value: null, provisional: true, interpretation: { headline: "Valuation history too thin", tone: "neutral", detail: "Fewer than 4 P/E points; provisional on the free tier." } };
  }
  const cheapSelf = 100 - pePctile;
  const peerAdj = Math.max(-15, Math.min(15, -0.3 * (peVsSector ?? 0)));
  const value = round(clamp(cheapSelf * 0.7 + 50 * 0.3 + peerAdj), 0);
  const sectorTxt = peVsSector == null ? "" : ` ${peVsSector >= 0 ? "+" : "−"}${Math.abs(peVsSector)}% vs sector;`;
  const tone: Interpretation["tone"] = pePctile <= 40 ? "good" : pePctile >= 75 ? "warn" : "neutral";
  const room = pePctile <= 40 ? "re-rating room" : pePctile >= 75 ? "priced for growth" : "fairly valued vs its own range";
  return { value, provisional: true, interpretation: { headline: `P/E in the ${pePctile}th percentile of its own range;${sectorTxt} ${room}`, tone, detail: "Per-name valuation is sparse on the free tier — provisional." } };
}

const legLabel: Record<keyof typeof WEIGHTS, string> = { technical: "technicals", growth: "the growth line", ownership: "ownership", valuation: "valuation" };

export function buildComposite(input: { chart: StockChartResponse; fund: Fundamentals; sectorPeMedian: number | null }): CompositeScore {
  const { chart, fund, sectorPeMedian } = input;
  const technical = technicalLeg(chart);
  const growth = growthLeg(fund);
  const ownership = ownershipLeg(chart.symbol);
  const valuation = valuationLeg(fund, sectorPeMedian);
  const legs = { technical, growth, ownership, valuation };

  // Headline: full-weight provisional legs; drop only NULL legs and renormalize.
  let wSum = 0;
  let acc = 0;
  (Object.keys(WEIGHTS) as (keyof typeof WEIGHTS)[]).forEach((k) => {
    const v = legs[k].value;
    if (v != null) {
      wSum += WEIGHTS[k];
      acc += WEIGHTS[k] * v;
    }
  });
  const headline = wSum > 0 ? round(acc / wSum, 0) : 0;

  const t = technical.value ?? 0;
  const stage = chart.stage;
  let verdict: Verdict;
  if (stage === 4 || t < 35 || headline < 45) verdict = "avoid";
  else if (headline >= 68 && t >= 55 && stage === 2) verdict = "act"; // Stage-2 gate
  else verdict = "watch";

  return { headline, verdict, technical, growth, ownership, valuation, interpretation: verdictSentence(verdict, legs, stage) };
}

function verdictSentence(verdict: Verdict, legs: Record<keyof typeof WEIGHTS, SubScore>, stage: Stage | null): Interpretation {
  const scored = (Object.keys(WEIGHTS) as (keyof typeof WEIGHTS)[]).filter((k) => legs[k].value != null);
  const dominant = scored.slice().sort((a, b) => (legs[b].value as number) * WEIGHTS[b] - (legs[a].value as number) * WEIGHTS[a])[0];
  const drag = scored.slice().sort((a, b) => (legs[a].value as number) - (legs[b].value as number))[0];
  const tone: Interpretation["tone"] = verdict === "act" ? "good" : verdict === "avoid" ? "bad" : "warn";
  const word = verdict === "act" ? "Act" : verdict === "avoid" ? "Avoid" : "Watch";
  const dragProv = drag && legs[drag].provisional ? " (provisional)" : "";
  let detail: string;
  if (verdict === "avoid") {
    detail = stage === 4 ? "In a Stage-4 downtrend — stay away regardless of the numbers." : `${cap(legLabel[drag])} drag the case${dragProv}.`;
  } else if (verdict === "act") {
    detail = `${cap(legLabel[dominant])} lead; ${legLabel[drag]} the only soft spot${dragProv}.`;
  } else {
    detail = `${cap(legLabel[dominant])} support it but ${legLabel[drag]} hold it back${dragProv} — watch, don't chase.`;
  }
  const clause = verdict === "act" ? "the setup lines up" : verdict === "avoid" ? "the case is weak" : "it's mixed";
  return { headline: `${word} — ${clause}`, tone, detail };
}

function cap(s: string): string {
  return s.charAt(0).toUpperCase() + s.slice(1);
}

File: api/composite.py
"""TechnoFunda composite score — Python mirror of the web's composite.ts, fed
from stored technicals + fundamentals + ownership (IBD50) + valuation
(re-rating). Provisional legs count at full weight; only null legs drop and the
rest renormalise."""
from __future__ import annotations

from .interpret import STAGE_SHORT, STAGE_TONE, interp

FLAT_PCT = 1.5
WEIGHTS = {"technical": 0.40, "growth": 0.30, "ownership": 0.15, "valuation": 0.15}


def _clamp(x: float) -> float:
    return max(0.0, min(100.0, x))


def _sub(value, provisional, interpretation) -> dict:
    return {"value": (None if value is None else round(value)), "provisional": provisional, "interpretation": interpretation}


def technical_leg(stage, range_pos, rs_new_high, above_ma, slope_pct) -> dict:
    if stage is None:
        return _sub(None, False, interp("No technical read", "neutral", "Needs 170 sessions of history."))
    base = {1: 45, 2: 75, 3: 35, 4: 12}[stage]
    proximity = 0.15 * ((range_pos if range_pos is not None else 50) - 50)
    rs_bonus = 12 if rs_new_high else 0
    trend_bonus = 6 if above_ma else 0
    slope_nudge = 0 if slope_pct is None else (4 if slope_pct > FLAT_PCT else -4 if slope_pct < -FLAT_PCT else 0)
    val = _clamp(base + proximity + rs_bonus + trend_bonus + slope_nudge)
    rp = range_pos if range_pos is not None else 50
    range_word = "near the top of its 52-week range" if rp >= 66 else "near the low of its range" if rp <= 33 else "mid-range"
    rs_word = "RS at a new high (leading)" if rs_new_high else "RS not yet leading"
    return _sub(val, False, interp(STAGE_SHORT[stage], STAGE_TONE[stage], f"{range_word}; {rs_word}."))


def growth_leg(rev_yoy, eps_yoy) -> dict:
    if rev_yoy is None and eps_yoy is None:
        return _sub(None, True, interp("Growth history unavailable", "neutral", "Not enough reported periods."))
    parts = []
    if rev_yoy is not None:
        parts.append(_clamp(50 + 1.4 * rev_yoy))
    if eps_yoy is not None:
        parts.append(_clamp(50 + 1.0 * eps_yoy))
    base = sum(parts) / len(parts)
    if eps_yoy is not None and eps_yoy < 0:
        base -= 10
    val = _clamp(base)
    tone = "good" if val >= 62 else "bad" if val < 45 else "neutral"

    def t(x):
        return "n/a" if x is None else f"{'+' if x >= 0 else '−'}{abs(x):.0f}%"

    return _sub(val, rev_yoy is None or eps_yoy is None, interp(f"Revenue {t(rev_yoy)} YoY, EPS {t(eps_yoy)}", tone, None))


def ownership_leg(is_ibd50: bool) -> dict:
    if is_ibd50:
        return _sub(78, True, interp("In the IBD 50", "good", "Institutional sponsorship present; provisional until a real 13F feed lands."))
    return _sub(46, True, interp("Not in the IBD 50", "neutral", "Thin ownership signal; provisional until a real 13F feed lands."))


def valuation_leg(pe_pctile, pe_vs_sector) -> dict:
    if pe_pctile is None:
        return _sub(None, True, interp("Valuation history too thin", "neutral", "Fewer than 4 P/E points; provisional on the free tier."))
    cheap_self = 100 - pe_pctile
    peer_adj = max(-15.0, min(15.0, -0.3 * (pe_vs_sector or 0)))
    val = _clamp(cheap_self * 0.7 + 50 * 0.3 + peer_adj)
    tone = "good" if pe_pctile <= 40 else "warn" if pe_pctile >= 75 else "neutral"
    room = "re-rating room" if pe_pctile <= 40 else "priced for growth" if pe_pctile >= 75 else "fairly valued vs its own range"
    vs = "" if pe_vs_sector is None else f" {'+' if pe_vs_sector >= 0 else '−'}{abs(round(pe_vs_sector))}% vs sector;"
    return _sub(val, True, interp(f"P/E in the {round(pe_pctile)}th percentile of its own range;{vs} {room}", tone, "Per-name valuation is sparse on the free tier — provisional."))


_LEG_LABEL = {"technical": "technicals", "growth": "the growth line", "ownership": "ownership", "valuation": "valuation"}


def build_composite(*, stage, range_pos, rs_new_high, above_ma, slope_pct, rev_yoy, eps_yoy, is_ibd50, pe_pctile, pe_vs_sector) -> dict:
    legs = {
        "technical": technical_leg(stage, range_pos, rs_new_high, above_ma, slope_pct),
        "growth": growth_leg(rev_yoy, eps_yoy),
        "ownership": ownership_leg(is_ibd50),
        "valuation": valuation_leg(pe_pctile, pe_vs_sector),
    }
    wsum = acc = 0.0
    for k, w in WEIGHTS.items():
        v = legs[k]["value"]
        if v is not None:
            wsum += w
            acc += w * v
    headline = round(acc / wsum) if wsum > 0 else 0
    t = legs["technical"]["value"] or 0
    if stage == 4 or t < 35 or headline < 45:
        verdict = "avoid"
    elif headline >= 68 and t >= 55 and stage == 2:
        verdict = "act"
    else:
        verdict = "watch"
    return {**legs, "headline": headline, "verdict": verdict, "interpretation": _verdict_sentence(verdict, legs, stage)}


def _verdict_sentence(verdict, legs, stage) -> dict:
    scored = [k for k in WEIGHTS if legs[k]["value"] is not None]
    dominant = max(scored, key=lambda k: legs[k]["value"] * WEIGHTS[k]) if scored else None
    drag = min(scored, key=lambda k: legs[k]["value"]) if scored else None
    tone = {"act": "good", "avoid": "bad", "watch": "warn"}[verdict]
    word = {"act": "Act", "avoid": "Avoid", "watch": "Watch"}[verdict]
    drag_prov = " (provisional)" if drag and legs[drag]["provisional"] else ""
    if verdict == "avoid":
        detail = "In a Stage-4 downtrend — stay away regardless of the numbers." if stage == 4 else f"{_LEG_LABEL[drag].capitalize()} drag the case{drag_prov}."
    elif verdict == "act":
        detail = f"{_LEG_LABEL[dominant].capitalize()} lead; {_LEG_LABEL[drag]} the only soft spot{drag_prov}."
    else:
        detail = f"{_LEG_LABEL[dominant].capitalize()} support it but {_LEG_LABEL[drag]} hold it back{drag_prov} — watch, don't chase."
    clause = "the setup lines up" if verdict == "act" else "the case is weak" if verdict == "avoid" else "it's mixed"
    return interp(f"{word} — {clause}", tone, detail)
